node/node.go

```diff
@@ -0,0 +1,291 @@
+	stack   map[string]ServiceConstructor // Protocol stack registered into this node
+	emux    *event.TypeMux                // Event multiplexer used between the services of a stack
+
+	running  *p2p.Server        // Currently running P2P networking layer
+	services map[string]Service // Currently running services
+
+	stop chan struct{} // Channel to wait for termination notifications
+	lock sync.RWMutex
+}
+
+// New creates a new P2P node, ready for protocol registration.
+func New(conf *Config) (*Node, error) {
+	// Ensure the data directory exists, failing if it cannot be created
+	if conf.DataDir != "" {
+		if err := os.MkdirAll(conf.DataDir, 0700); err != nil {
+			return nil, err
+		}
+	}
+	// Assemble the networking layer and the node itself
+	nodeDbPath := ""
+	if conf.DataDir != "" {
+		nodeDbPath = filepath.Join(conf.DataDir, datadirNodeDatabase)
+	}
+	return &Node{
+		datadir: conf.DataDir,
+		config: &p2p.Server{
+			PrivateKey:      conf.NodeKey(),
+			Name:            conf.Name,
+			Discovery:       !conf.NoDiscovery,
+			BootstrapNodes:  conf.BootstrapNodes,
+			StaticNodes:     conf.StaticNodes(),
+			TrustedNodes:    conf.TrusterNodes(),
+			NodeDatabase:    nodeDbPath,
+			ListenAddr:      conf.ListenAddr,
+			NAT:             conf.NAT,
+			Dialer:          conf.Dialer,
+			NoDial:          conf.NoDial,
+			MaxPeers:        conf.MaxPeers,
+			MaxPendingPeers: conf.MaxPendingPeers,
+		},
+		stack: make(map[string]ServiceConstructor),
+		emux:  new(event.TypeMux),
+	}, nil
+}
+
+// Register injects a new service into the node's stack.
+func (n *Node) Register(id string, constructor ServiceConstructor) error {
+	n.lock.Lock()
+	defer n.lock.Unlock()
+
+	// Short circuit if the node is running or if the id is taken
+	if n.running != nil {
+		return ErrNodeRunning
+	}
+	if _, ok := n.stack[id]; ok {
+		return ErrServiceRegistered
+	}
+	// Otherwise register the service and return
+	n.stack[id] = constructor
+
+	return nil
+}
+
+// Unregister removes a service from a node's stack. If the node is currently
+// running, an error will be returned.
+func (n *Node) Unregister(id string) error {
+	n.lock.Lock()
+	defer n.lock.Unlock()
+
+	// Short circuit if the node is running, or if the service is unknown
+	if n.running != nil {
+		return ErrNodeRunning
+	}
+	if _, ok := n.stack[id]; !ok {
+		return ErrServiceUnknown
+	}
+	// Otherwise drop the service and return
+	delete(n.stack, id)
+	return nil
+}
+
+// Start create a live P2P node and starts running it.
+func (n *Node) Start() error {
+	n.lock.Lock()
+	defer n.lock.Unlock()
+
+	// Short circuit if the node's already running
+	if n.running != nil {
+		return ErrNodeRunning
+	}
+	// Otherwise copy and specialize the P2P configuration
+	running := new(p2p.Server)
+	*running = *n.config
+
+	ctx := &ServiceContext{
+		dataDir:  n.datadir,
+		EventMux: n.emux,
+	}
+	services := make(map[string]Service)
+	for id, constructor := range n.stack {
```

Comment:
Ok, listen address is a valid use case. I'll consider how to cleanly add them.

There's a more subtle issue in there too. The list of services to instantiate is a map, not a slice, so there are no ordering guarantees. This means that a constructor cannot assume that some dependent service was already launched. I'll rework to retain the registration order when constructing.